server/util/perms/perms_test.go

```diff
@@ -0,0 +1,313 @@
+		{
+			name: "missing group perms",
+			acl: &aclpb.ACL{
+				UserId:            &uidpb.UserId{Id: "US1"},
+				GroupId:           "GR1",
+				OwnerPermissions:  &aclpb.ACL_Permissions{Read: true},
+				GroupPermissions:  nil,
+				OthersPermissions: &aclpb.ACL_Permissions{Read: true},
+			},
+			wantErr: status.InvalidArgumentError("ACL is missing one or more required permissions fields."),
+		},
+		{
+			name: "missing others perms",
+			acl: &aclpb.ACL{
+				UserId:            &uidpb.UserId{Id: "US1"},
+				GroupId:           "GR1",
+				OwnerPermissions:  &aclpb.ACL_Permissions{Read: true},
+				GroupPermissions:  &aclpb.ACL_Permissions{Read: true},
+				OthersPermissions: nil,
+			},
+			wantErr: status.InvalidArgumentError("ACL is missing one or more required permissions fields."),
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := perms.FromACL(test.acl)
+			require.Equal(t, test.wantErr, err)
+		})
+	}
+}
+
+func TestAuthorizeRead(t *testing.T) {
+	for _, test := range []struct {
+		name                string
+		authenticatedUserID string
+		acl                 *aclpb.ACL
+		anonAuthDisabled    bool
+		wantErr             error
+		wantAnonymousErr    bool
+	}{
+		{
+			name:    "nil user",
+			acl:     &aclpb.ACL{},
+			wantErr: status.InvalidArgumentError("user cannot be nil."),
+		},
+		{
+			name:                "invalid ACL",
+			authenticatedUserID: "US1",
+			acl: &aclpb.ACL{
+				UserId:            &uidpb.UserId{Id: "US1"},
+				GroupPermissions:  &aclpb.ACL_Permissions{Read: true},
+				OthersPermissions: &aclpb.ACL_Permissions{Read: true},
+			},
+			wantErr: status.InvalidArgumentError("ACL is missing one or more required permissions fields."),
+		},
+		{
+			name:                "authenticated user with owner read perms",
+			authenticatedUserID: "US1",
+			acl: &aclpb.ACL{
+				UserId:            &uidpb.UserId{Id: "US1"},
+				OwnerPermissions:  &aclpb.ACL_Permissions{Read: true},
+				GroupPermissions:  &aclpb.ACL_Permissions{},
+				OthersPermissions: &aclpb.ACL_Permissions{},
+			},
+			wantErr: nil,
+		},
+		{
+			name:                "authenticated user with group read perms",
+			authenticatedUserID: "US2",
+			acl: &aclpb.ACL{
+				UserId:            &uidpb.UserId{Id: "US1"},
+				GroupId:           "GR1",
+				OwnerPermissions:  &aclpb.ACL_Permissions{Read: true},
+				GroupPermissions:  &aclpb.ACL_Permissions{Read: true},
+				OthersPermissions: &aclpb.ACL_Permissions{},
+			},
+			wantErr: nil,
+		},
+		{
+			name:                "authenticated user without owner or group read perms",
+			authenticatedUserID: "US1",
+			acl: &aclpb.ACL{
+				UserId:            &uidpb.UserId{Id: "US3"},
+				GroupId:           "GR3",
+				OwnerPermissions:  &aclpb.ACL_Permissions{Read: true},
+				GroupPermissions:  &aclpb.ACL_Permissions{Read: true},
+				OthersPermissions: &aclpb.ACL_Permissions{},
+			},
+			wantErr: status.PermissionDeniedError("You do not have permission to perform this action."),
+		},
+	} {
```

Comment:
nit: could we add a cross-group read test (GR2 attempting to read GR3 data)?

Reply:
I think that is already being tested in `authenticated user without owner or group read perms`? The user US1 is a member of GR1 and trying to read an object owned by GR3